Wabbajack.Networking.Http/ResumableDownloader.cs

```diff
@@ -1,187 +1,146 @@
-    }
-
-    public async Task<Hash> Download(CancellationToken token)
-    {
-        _token = token;
-
-        var downloader = new DownloadService(CreateConfiguration(_msg));
-        downloader.DownloadStarted += OnDownloadStarted;
-        downloader.DownloadProgressChanged += OnDownloadProgressChanged;
-        downloader.DownloadFileCompleted += OnDownloadFileCompleted;
-
-        // Attempt to resume previous download
-        var downloadPackage = LoadPackage();
-        if (downloadPackage != null)
+        if (_msg.RequestUri == null)
         {
-            // Resume with different Uri in case old one is no longer valid
-            downloadPackage.Address = _msg.RequestUri!.AbsoluteUri;
+            throw new ArgumentException("Request URI is null");
+        }
 
-            _logger.LogDebug("Download for {name} is resuming...", _outputPath.FileName.ToString());
-            await downloader.DownloadFileTaskAsync(downloadPackage, token);
+        try
+        {
+            return await DownloadAndHash(_msg, _outputPath, job, token, 5);
         }
-        else
+        catch (Exception ex)
         {
-            _logger.LogDebug("Download for '{name}' is starting from scratch...", _outputPath.FileName.ToString());
-            _outputPath.Delete();
-            await downloader.DownloadFileTaskAsync(_msg.RequestUri!.AbsoluteUri, _outputPath.ToString(), token);
+            _logger.LogError(ex, "Failed to download '{name}'", _outputPath.FileName.ToString());
+
+            throw;
         }
+    }
 
-        // Save progress if download isn't completed yet
-        if (downloader.Status is DownloadStatus.Stopped or DownloadStatus.Failed)
+    private async Task<Hash> DownloadAndHash(HttpRequestMessage msg, AbsolutePath filePath, IJob job, CancellationToken token, int retry = 5, bool reset = false)
+    {
+        try
         {
-            _logger.LogDebug("Download for '{name}' stopped before completion. Saving package...", _outputPath.FileName.ToString());
-            SavePackage(downloader.Package);
-            if (_error == null || _error.GetType() == typeof(TaskCanceledException))
+            if (reset)
             {
-                return new Hash();
+                filePath.Delete();
             }
 
-            if (_error.GetType() == typeof(NotSupportedException))
+            var downloadedFilePath = await DownloadStreamDirectlyToFile(msg, filePath, job, token);
+
+            return await HashFile(downloadedFilePath, token);
+        }
+        catch (Exception ex) when (ex is SocketException || ex is IOException)
+        {
+            _logger.LogWarning(ex, "Failed to download '{name}' due to network error. Retrying...", filePath.FileName.ToString());
+
+            if (retry == 0)
             {
-                _logger.LogWarning("Download for '{name}' doesn't support resuming. Deleting package...", _outputPath.FileName.ToString());
-                DeletePackage();
+                throw;
```

Comment:
Would be nice to keep a log message here so we're sure it threw in this spot

Reply:
Added